share/doc/wake/from-scala.adoc

```diff
@@ -0,0 +1,401 @@
+----
+
+
+| Get or else
+|
+[source,scala]
+----
+val foo = opt.getOrElse(2)
+----
+|
+[source,wake]
+----
+def foo = opt \| getOrElse 2
+----
+
+
+| Testing for Some/None
+|
+[source,scala]
+----
+val foo = opt.isDefined
+val bar = opt.isEmpty
+----
+|
+[source,wake]
+----
+def foo = opt.isSome
+def bar = opt.isNone
+----
+
+|===
+
+
+== List
+
+Both Scala and Wake have a list type named `List`, corresponding to a linked-list structure.
+Unlike Scala, Wake does not define a constructor with a variable number of arguments.
+Instead, it is idiomatic in Wake to define lists using its "cons" operator `,`.
+
+[[table-list]]
+.List type in Scala and Wake
+[%autowidth.stretch,cols=",a,a"]
+|===
+| Description | Scala | Wake
+
+
+| Empty list
+|
+[source,scala]
+----
+Nil
+----
+|
+[source,wake]
+----
+Nil
+----
+
+
+| Defining a list
+|
+[source,scala]
+----
+List(1, 2, 3)
+// Or equivalently
+1 :: 2 :: 3 :: Nil
+----
+|
+[source,wake]
+----
+1, 2, 3, Nil
+----
+
+
+| Map
+|
+[source,scala]
+----
+val bar = foo.map(_ + 1)
+----
+|
+[source,wake]
+----
+def bar = foo \| map (_ + 1)
+----
+
+
+| Flat map
+|
+[source,scala]
+----
+val words = List("foo", "bar")
+val chars = words.flatMap(_.toList)
+----
+|
+[source,wake]
+----
+def words = "foo", "bar", Nil
+def chars = words \| mapFlat explode
+----
+
+|===
+
+
+== Try or Either
+
+A very important data type in Wake is `Result`, representing the success or failure of an action.
+It is equivalent to Scala's `Try` (or Scala's `Either`), with `Result a b` describing an object that can either be a `Pass a` or a `Fail b`.
+However, because Wake is designed to describe the execution of build commands, `Result` objects are ubiquitous in Wake code.
+
+In addition Wake defines an `Error` type, consisting of a `Cause` string and a `Stack` trace, and it is common for the `Fail` type parameter to be an `Error` type.
+
+See <<table-result>> for more information.
+
+[[table-result]]
+.Scala's Try type compared to Wake's Result type
+[%autowidth.stretch,cols=",a,a"]
+|===
+| Description | Scala | Wake
+
+| Constructor
+|
+[source,scala]
+----
+Success(1)
+Failure(exceptionObject)
+----
+|
+[source,wake]
+----
+Pass 1
+Fail exceptionObject
+# Construct an Error object with stack trace
+Fail (makeError "My error")
+----
+
+
+| Mapping on successful case
+|
+[source,scala]
+----
+val result = Success(1)
+val newResult = result.map(_ + 1)
+----
+|
+[source,wake]
+----
+def result = Pass 1
+def newResult = result \| rmap (_ + 1)
+----
+
+
+| Checking for success/failure
+|
+[source,scala]
+----
+result.isSuccess
+result.isFailure
+----
+|
+[source,wake]
+----
+result.isPass
+result.isFail
+----
+
+
+| Flat map
+|
+[source,scala]
+----
+// Could fail if d == 0
+def reciprocal(d: Double): Try[Double] = Try { 1.0 / d }
+
+val result = Success(1)
+val newResult = result.flatMap(reciprocal)
+----
+|
+[source,wake]
+----
+def reciprocal d =
+  if (d == 0.0)
+    makeError("Divide by zero")
+  else
+    1.0 / d
```

Comment:
You presumably need a Pass + Fail here. You also need to decide if it's a Double or an Integer. How about?:
```
  def reciprocal = match _
    0.0 = Fail (makeError "Divide by Zero")
    d = Pass (1.0 /. d)
```